Foreman/DataCache.cs

```diff
@@ -1084,58 +1093,108 @@ private static Item FindOrCreateUnknownItem(String itemName)
         private static Bitmap GetIcon(LuaTable values, bool fallbackToUnknown = false)
         {
             string fileName = ReadLuaString(values, "icon", true);
-            if (fileName == null)
+            string mipmaps = ReadLuaString(values, "icon_mipmaps", true);
+            int iconsize = 0;
+            if (mipmaps != null)
             {
-                LuaTable icons = ReadLuaLuaTable(values, "icons", true);
+                iconsize = 4 * Convert.ToInt32(Math.Pow(2, Convert.ToDouble(mipmaps)));
+            }
 
-                try
+
+
+            return (fileName == null ? GetIcons(values) : LoadImage(fileName, iconsize));
+        }
+        private static Bitmap GetIcons(LuaTable values)
+        {
+            Bitmap fullImage = new Bitmap(32,32);
+            for (int i = 0; i < 32; i++)
+            {
+                for (int j = 0; j < 32; j++)
                 {
-                    if (!Convert.ToString(values["name"]).Contains("barrel"))
+                    fullImage.SetPixel(i, j, Color.FromArgb(0, 255, 255, 255));
+                }
+            }
+            LuaTable icons = ReadLuaLuaTable(values, "icons", true);
+            try
+            {
+                if (!Convert.ToString(values["name"]).Contains("barrel"))
+                {
+                    foreach (var iconKey in icons.Keys)
                     {
-                        //ErrorLogging.LogLine("recipe name: " + values["name"]);
-                        foreach (var s in icons.Values)
+                        if (iconKey != null)

```

Comment:
Can we combine this checks into one lines please? And if "false" can we log an error?

Reply:
already done in my local repo like this:
`if (iconKey != null && icons[iconKey] is LuaTable)`